core/src/trezor/crypto/slip39.py

```diff
@@ -0,0 +1,606 @@
+    """Converts a list of base 1024 indices in big endian order to an integer value."""
+    value = 0
+    for index in indices:
+        value = (value << RADIX_BITS) + index
+    return value
+
+
+def _int_to_indices(value, length, bits):
+    """Converts an integer value to indices in big endian order."""
+    mask = (1 << bits) - 1
+    return ((value >> (i * bits)) & mask for i in reversed(range(length)))
+
+
+def mnemonic_from_indices(indices):
+    return " ".join(wordlist[i] for i in indices)
+
+
+def mnemonic_to_indices(mnemonic):
+    return (word_index(word.lower()) for word in mnemonic.split())
+
+
+def _round_function(i, passphrase, e, salt, r):
+    """The round function used internally by the Feistel cipher."""
+    return pbkdf2(
+        pbkdf2.HMAC_SHA256,
+        bytes([i]) + passphrase,
+        salt + r,
+        (BASE_ITERATION_COUNT << e) // ROUND_COUNT,
+    ).key()[: len(r)]
+
+
+def _get_salt(identifier):
+    return CUSTOMIZATION_STRING + identifier.to_bytes(
+        bits_to_bytes(ID_LENGTH_BITS), "big"
+    )
+
+
+def _encrypt(master_secret, passphrase, iteration_exponent, identifier):
+    l = master_secret[: len(master_secret) // 2]
+    r = master_secret[len(master_secret) // 2 :]
+    salt = _get_salt(identifier)
+    for i in range(ROUND_COUNT):
+        (l, r) = (
+            r,
+            xor(l, _round_function(i, passphrase, iteration_exponent, salt, r)),
+        )
+    return r + l
+
+
+def decrypt(identifier, iteration_exponent, encrypted_master_secret, passphrase):
+    l = encrypted_master_secret[: len(encrypted_master_secret) // 2]
+    r = encrypted_master_secret[len(encrypted_master_secret) // 2 :]
+    salt = _get_salt(identifier)
+    for i in reversed(range(ROUND_COUNT)):
+        (l, r) = (
+            r,
+            xor(l, _round_function(i, passphrase, iteration_exponent, salt, r)),
+        )
+    return r + l
+
+
+def _create_digest(random_data, shared_secret):
+    return hmac.new(random_data, shared_secret, hashlib.sha256).digest()[
+        :DIGEST_LENGTH_BYTES
+    ]
+
+
+def _split_secret(threshold, share_count, shared_secret):
+    assert 0 < threshold <= share_count <= MAX_SHARE_COUNT
+
+    # If the threshold is 1, then the digest of the shared secret is not used.
+    if threshold == 1:
+        return [(i, shared_secret) for i in range(share_count)]
+
+    random_share_count = threshold - 2
+
+    if share_count > MAX_SHARE_COUNT:
+        raise ValueError(
+            "The requested number of shares ({}) must not exceed {}.".format(
+                share_count, MAX_SHARE_COUNT
+            )
+        )
+
+    shares = [(i, random.bytes(len(shared_secret))) for i in range(random_share_count)]
+
+    random_part = random.bytes(len(shared_secret) - DIGEST_LENGTH_BYTES)
+    digest = _create_digest(random_part, shared_secret)
+
+    base_shares = shares + [
+        (DIGEST_INDEX, digest + random_part),
+        (SECRET_INDEX, shared_secret),
+    ]
+
```

Comment:
I suggest checking that all the shares have equal length, even though it is done in `modtrezorcrypto-shamir.h`.